File: examples/intermediate/lqr_pendulum.cpp
#include "prx/utilities/defs.hpp"
#include "prx/planning/world_model.hpp"
#include "prx/simulation/plants/plants.hpp"
#include "prx/planning/condition_check.hpp"
#include "prx/simulation/controllers/lqr.hpp"
#include "prx/visualization/three_js_group.hpp"
#include "prx/utilities/general/param_loader.hpp"
#include "prx/simulation/loaders/obstacle_loader.hpp"

using namespace prx;

int main(int argc, char* argv[])
{
    auto params = param_loader("examples/intermediate/lqr.yaml", argc, argv);

	simulation_step = params["simulation_step"].as<double>();
    init_random(params["random_seed"].as<int>());

    auto obstacles = load_obstacles(params["environment"].as<>());
    std::vector<std::shared_ptr<movable_object_t>> obstacle_list = obstacles.second;
    std::vector<std::string> obstacle_names = obstacles.first;

    std::string plant_name = params["/plant/name"].as<>();
    std::string plant_path = params["/plant/path"].as<>();
    auto plant = prx::system_factory_t::create_system(plant_name, plant_path);
    prx_assert(plant != nullptr, "Plant is nullptr!");

    world_model_t<> world_model({plant},{obstacle_list});
    world_model.create_context("dirt_context",{plant_name},{obstacle_names});
    auto context = world_model.get_context("dirt_context");

    const auto ss = context.first -> get_state_space();
    const auto cs = context.first -> get_control_space();

    auto lower_bounds = params["/plant/state_space_lower_bound"].as<std::vector<double>>();
    auto upper_bounds = params["/plant/state_space_upper_bound"].as<std::vector<double>>();
    ss -> set_bounds(lower_bounds, upper_bounds);

    auto start_state = ss -> make_point();
	auto goal_state = ss -> make_point();

    ss -> copy_point_from_vector(start_state, params["/plant/start_state"].as<std::vector<double>>());
    ss -> copy_point_from_vector(goal_state, params["/plant/goal_state"].as<std::vector<double>>());
    ss -> copy_from_point(start_state);
    
    condition_check_t checker(params["checker_type"].as<>(), params["checker_value"].as<int>());

    trajectory_t solution_traj(ss);

    auto pendulum = std::dynamic_pointer_cast<prx::pendulum_t>(plant);
    pendulum -> linearize();


    auto Q = Eigen::MatrixXd::Identity(2,2);
    auto R = Eigen::MatrixXd::Identity(1,1);
    lqr_t lqr(pendulum, Q, R, "LQR");
    lqr.compute_K();
    Eigen::MatrixXd K = lqr.get_K();
    std::cout << "K: " << K << std::endl;
    
    do
    {
        cs -> enforce_bounds();
        lqr.compute_controls();
        plant -> propagate(simulation_step);
        std::cout << "[pendulum] " << plant << std::endl;
        solution_traj.copy_onto_back(ss);

    }
    while(!checker.check() && space_t::euclidean_2d(solution_traj.back(), goal_state, 0, 2) > 0.01);

    three_js_group_t* vis_group = new three_js_group_t({plant},{obstacle_list});

    std::string body_name = params["/plant/name"].as<>() + "/" + params["/plant/vis_body"].as<>();

    vis_group -> add_detailed_vis_infos(info_geometry_t::FULL_LINE, solution_traj, 
        body_name, ss);

    vis_group -> add_animation(solution_traj, ss, start_state);

    vis_group -> output_html("lqr_pendulum.html");

    delete vis_group;

    params.print();
}

File: examples/tripods/lqr_pendulum.py
import sys
import math
import random
# Remember to add libpyDirtMP to your PYTHONPATH
# On bash: ``export PYTHONPATH=$DIRTMP_PATH/lib/:$PYTHONPATH
import libpyDirtMP as prx 

if __name__ == "__main__":

	# This is able to handle *.yaml files as well as command line arguments
	# in the form of --argument=value and --/arg_root/arg_l1/arg_l2=value
	# i.e.: 
	# 	Simple example: python lqr_pendulum.py --checker_value=5  
	# 	Complex (with vectors!): python lqr_pendulum.py --/plant/start_state=\[1,-1.0]
	# 	More complex: python lqr_pendulum.py --checker_value=5  --/plant/start_state=\[1,-1.0]
	params = prx.param_loader("examples/intermediate/lqr.yaml", sys.argv);
	print(params.get_input_path())
	params.print()

	prx.set_simulation_step(params["simulation_step"].as_float())
	prx.init_random(params["random_seed"].as_int())

	# Could be done directly from params, calling the getter to show that is there...
	simulation_step = prx.get_simulation_step()
	# Safety check
	print("simulation_step:", simulation_step)

	obstacles = prx.load_obstacles(params["environment"].as_string())
	obstacle_list = obstacles.objects
	obstacle_names = obstacles.names

	plant_name = params["/plant/name"].as_string()
	plant_path = params["/plant/path"].as_string()
	plant = prx.system_factory.create_system(plant_name, plant_path);
	# plant = prx.pendulum(plant_name, plant_path)
	# prx_assert(plant != nullptr, "Plant is nullptr!");
	if plant == None: 
		print("Error: plant not found!")
		exit(-1)

	wm = prx.world_model([plant], obstacle_list)
	wm.create_context("context", [plant_name], obstacle_names)
	context = wm.get_context("context")

	ss = context.system_group.get_state_space()
	cs = context.system_group.get_control_space()

	lower_bounds = params["/plant/state_space_lower_bound"].as_float_vector()
	upper_bounds = params["/plant/state_space_upper_bound"].as_float_vector()
	ss.set_bounds(lower_bounds, upper_bounds)

	cs_lb = params["/plant/control_space_lower_bound"].as_float_vector()
	cs_up = params["/plant/control_space_upper_bound"].as_float_vector()
	cs.set_bounds(cs_lb, cs_up);


	start_state = ss.make_point()
	goal_state  = ss.make_point()

	ss.copy_point_from_vector(start_state, params["/plant/start_state"].as_float_vector())
	ss.copy_point_from_vector(goal_state, params["/plant/goal_state"].as_float_vector())
	ss.copy_from_point(start_state)
	
	checker = prx.condition_check(params["checker_type"].as_string(), params["checker_value"].as_int())

	solution_traj = prx.trajectory(ss)

	solution_traj.copy_onto_back(ss)
	# plant.__class__ = prx.pendulum
	# print(repr(plant))
	plant.linearize()

	Q = prx.matrix.Identity(2,2)
	R = prx.matrix.Identity(1,1)
	print("Q:", Q)
	print("R:", R)

	lqr = prx.lqr(plant, Q, R, "LQR");
	lqr.compute_K();
	K = lqr.get_K();
	print("K:", K)
	# simulating a do{}while()
	print("start_state:", start_state)
	while True:
		lqr.compute_controls();
		cs.enforce_bounds();
		plant.propagate(simulation_step);
	# 	# std::cout << "[pendulum] " << plant << std::endl;
		solution_traj.copy_onto_back(ss);
		if checker.check():
			break

	vis_group = prx.three_js_group([plant], obstacle_list)

	body_name = params["/plant/name"].as_string() + "/" + params["/plant/vis_body"].as_string()

	vis_group.add_detailed_vis_infos(prx.info_geometry.FULL_LINE, solution_traj, body_name, ss, "0xFF0000")

	vis_group.add_animation(solution_traj, ss, start_state)

	vis_group.output_html("py_lqr_pendulum_control.html")